### PR TITLE
[v3] WIP: Make types compatible with TypeScript 2.4

### PR DESCRIPTION
**Pull request by lumaxis**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -3001,11 +3001,11 @@ declare module sequelize {
   }
 
   export interface OrOperator {
-    $or: WhereOperators | WhereAttributeHash | Array<WhereOperators | WhereAttributeHash>;
+    $or: WhereOperators | WhereAttributeHash | Array<Array<number> | WhereOperators | WhereAttributeHash>;
```
Comment (felixfbecker) [inline index.d.ts]: not sure what `$or: [[1, 2, 3]]` is supposed to mean?

Comment (lumaxis) [inline index.d.ts]: @felixfbecker I think this was necessary to cover this test case:
```typescript
id: {
  $or: [
    [1, 2, 3],
    { $gt: 10 }
  ]
}
```

Comment (felixfbecker) [inline index.d.ts]: So what is the `[1, 2, 3]` supposed to mean then? Do you mean `$in` or `$any`?

Comment (lumaxis) [inline index.d.ts]: @felixfbecker I just tried it and looks like Sequelize treats an array as an `$in` automatically and generates something like this:
```sql
SELECT [id], [created_at], [updated_at] FROM [User] AS [User] WHERE ([User].[id] IN (1, 2, 3) OR [User].[id] > 10
```

Comment (felixfbecker) [inline index.d.ts]: got it, but it should allow more than just `number` then

Comment (lumaxis) [inline index.d.ts]: @felixfbecker Anything besides `string` you can think of?

Comment (felixfbecker) [inline index.d.ts]: It could be anything, right? number, string, boolean, JSON, an array itself?

Comment (lumaxis) [inline index.d.ts]: Probably true, but tbh, this feels kinda out of scope for this PR as its primary goal is to restore compatibility with the current version of TS.
I'm open to taking a look at extending/fixing this further but I'd like to unblock my upgrade to TS 2.4 first.

Comment (felixfbecker) [inline index.d.ts]: It's of limited use if it only fixes it for TS 2.4 if the user is using the operator on a number, but not on e.g. strings. I would use the `any` type here then

File: tsconfig.json [hunk 2]
```diff
@@ -1,5 +1,8 @@
 {
   "compilerOptions": {
+    "lib": [
+      "es2015"
```
Comment (felixfbecker) [inline tsconfig.json]: Is this needed? v3 works with lower Node versions

Comment (lumaxis) [inline tsconfig.json]: Unfortunately, I think it is. The Bluebird types somewhere between 3.5.5 and 3.5.8 introduced the `Map` type which can't be handled by `tsc` with the standard settings. I added a bit of info to the commit message that added this config change: https://github.com/types/sequelize/pull/111/commits/0e75e9683ab71b1d25f9da28c543b3f32405a752

**Comment by lumaxis** [comment]
@felixfbecker Is there now anything you'd still like to see changed or are we good to merge?
